python: Add support for `black` formatter

mmacpherson opened the pull request:
This adds support for the [`black`](https://github.com/ambv/black) formatter to the python layer, and follows the existing YAPF code closely:

- Adds the [`blacken`](https://melpa.org/#/blacken) package
- Provides `blacken-buffer` as an alternative to `yapfify-buffer`
- Provisionally, binds `blacken-buffer` to "-", next to `yapfify-buffer`'s "="
- Adds `python-enable-black-format-on-save` variable
- Updates docs

Both `blacken-buffer` and the format-on-save behavior work for me on 25.2.2/ubuntu18.04.

tko commented:
I'm little concerned about trying to run two formatters in parallel. The `-format-on-save` hooks seem completely clashing with each other and consuming top-level keybinding for another formatter seems wasteful. (The latter would be less of an issue if formatters were under one or two more prefixes.)

While I can appreciate the desire for options (I'd like to have black) I think it should be a "pick one" choice (see html/javascript/json layers) and not something to be constantly thinking about with alternating keybindings. And if you're signing up for an opinionated code formatter it seems unlikely you'd go asking for second opinion from a different formatter all that often.

That said you may want to use different formatters for different projects with different conventions.

antonalekseev commented:
> While I can appreciate the desire for options (I'd like to have black) I think it should be a "pick one" choice (see html/javascript/json layers) and not something to be constantly thinking about with alternating keybindings.

I second that. Integration of `prettier` formatter in javascript/typescript layer (#10920, `javascript-fmt-tool`) may serve as an example.

mmacpherson commented:
@tko @antonalekseev Thank you, agreed. Had a go at abstracting out the choice of formatter in this latest commit, glad for any feedback. Continues to work for my setup described earlier, for both yapf and black.

One note in particular: here I removed the existing `python-enable-yapf-format-on-save` config variable in favor of the more general variables I introduced. That would break folks' existing configuration; if needed can imp backward compatibility.

verhovsky commented:
Can we get this merged?

voidlily commented:
Is this still getting merged?

smile13241324 commented:
Sorry lost track of it due to having just too many PRs open and forgot to assign to me. 

I put your PR on my task list and do it next time.

safijari commented:
I came looking for this just in time :sweat_smile:

smile13241324 commented:
Thank you for contributing to Spacemacs! :+1:
The PR has been cherry-picked into develop, you can safely delete your branch.

mmacpherson commented:
@smile13241324  Thank you. I've squashed the commits and pushed; please let me know if I might do anything further. Cheers.

smile13241324 commented:
Thank you for contributing to Spacemacs!
The PR has been cherry-picked into the develop branch, you can safely delete your branch.

casertap commented:
Thanks for this PR. How do I pass options to black?
I would like to pass the option `--exclude '/(\.eggs|\.git|\.hg|\.mypy_cache|\.nox|\.tox|\.venv|_build|buck-out|build|dist|\.requirements|\.serverless|venv)/' --line-length 120` for instance.
Thanks
